aesara/tensor/elemwise.py

```diff
@@ -1434,29 +1407,8 @@ def perform(self, node, inp, out):
             acc_dtype = node.outputs[0].type.dtype
 
         if to_reduce:
-            for dimension in to_reduce:
-                # If it's a zero-sized array, use scalar_op.identity
-                # if available
-                if variable.shape[dimension] == 0:
-                    if hasattr(self.scalar_op, "identity"):
-                        # Compute the shape of the output
-                        v_shape = list(variable.shape)
-                        del v_shape[dimension]
-                        variable = np.empty(tuple(v_shape), dtype=acc_dtype)
-                        variable.fill(self.scalar_op.identity)
-                    else:
-                        raise ValueError(
-                            f"Input ({variable}) has zero-size on axis {dimension}, but "
-                            f"self.scalar_op ({self.scalar_op}) has no attribute 'identity'"
-                        )
-                else:
-                    variable = self.ufunc.reduce(variable, dimension, dtype=acc_dtype)
-
-            variable = np.asarray(variable)
-            if np.may_share_memory(variable, input):
-                # perhaps numpy is clever for reductions of size 1?
-                # We don't want this.
-                variable = variable.copy()
+            variable = np.array(variable, dtype=acc_dtype)
+            variable = self.ufunc.reduce(variable, axis=tuple(axis))
```

Comment:
Since Numpy  version = 1.7.0, `ufunc.reduce` supports tuples in `axis` argument allowing reductions along multiple axes. This effectively makes something like `np.add.reduce()` equivalent to `np.sum()` . See [here](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.reduce.html) for details.

Reply:
Exactly; that was another one of the reasons for wanting to simplify this `Op`. I came across this during the implementation of the JAX conversion functions, and it's related to the `nfunc_variadic` work-around.

Note: the whole `Elemwise` + `scalar_op` thing might seem superfluous, but it does have a purpose&mdash;although not necessarily for computations (e.g. `Op.perform`).  It's used by the "fusion" optimization, which takes this information from multiple summed/multiplied `Elemwise`s and creates a single `Composite` `Op`.